### PR TITLE
[Bug-fix]: Windows VM HyperV Reenlightenment enabled fails to migrate

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
## Bug description & overview
A windows VM using HyperV Reenlightenment cannot migrate successfully. An example for such VM:
```yaml
spec:
  domain:
    features:
      hyperv:
        reenlightenment:
          enabled: true
```

(of course, the spec is trimmed down only to the relevant portion.

## Root cause
On [a commit to QEMU](https://gitlab.com/qemu-project/qemu/-/commit/561dbb41b1d752098249128d8462aaadc56fd15d), it says: `"Require 'tsc-frequency=' command line option to be specified for successful migration when re-enlightenment was enabled by the guest."`

IOW, the `tsc-frequency` parameter needs to pass to QEMU if HyperV Reenlightenment is enabled.

In converter.go [1], we're adding a tsc frequency parameter if `vmi.Status.TopologyHints != nil && vmi.Status.TopologyHints.TSCFrequency != nil`.
In virt-controller [2] we are checking if TopologyHints need to be added to the vmi by calling `TopologyHintsRequiredForVMI()`. There, we are checking if "invtsc" feature exists in vmi's spec (vmi.Spec.Domain.CPU.Features) with a "required" or "force" policy. Only then we add the TopologyHints to the vmi, which will lead (in [1]) to adding this parameter and pass it on to libvirt / QEMU.

[1] https://github.com/kubevirt/kubevirt/blob/v0.54.0/pkg/virt-launcher/virtwrap/converter/converter.go#L1691
[2] https://github.com/kubevirt/kubevirt/blob/v0.54.0/pkg/virt-controller/watch/vmi.go#L438

The reason this error popped up just now is that we bumped up libvirt & QEMU versions (from QEMU 5.2.0 to 6.2.0).

## Fix
As explained above, TopologyHints are added to a vmi if it has the `invtsc` feature and this will eventually cause us to provide 'tsc-frequency=' flag for QEMU [1].

<s>After this PR, invtsc is also considered as a hyperv EVMCS dependency. It's being added automatically via mutating webhooks and validated by validating webhooks.</s>

After this PR, TopologyHints will be added if either `invtsc` is enabled for the vmi (as before) or if the VMI is using HyperV Reenlightenment feature. In other words, after this PR there wouldn't be a need of any user intervention in order to migrate VMs with HyperV Reenlightenment enabled and the lower TSC frequency on the cluster would be automatically set.

## Workaround
<s>Since currently 'tsc-frequency=' flag is passed to QEMU if invtsc feature is enabled, simply add the invtsc feature to the VM's spec:

```yaml
cpu:
  features:
  - name: invtsc
    policy: require
```

After doing so the VM is able to migrate successfully.
</s>

Workaround is not valid since KVM does not support windows + invtsc.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=2100054
Bugzilla (KBase): https://bugzilla.redhat.com/show_bug.cgi?id=2100629

**Notes for reviewers**
1. Added small refactoring to `hyperv.go`
2. Removed unnecessary call to `TopologyHintsRequiredForVMI()` on virt-controller's `vmi.go`

**Release note**:

```release-note
[Bug-fix]: Windows VM with WSL2 guest fails to migrate
```

### CODE REVIEW

**Comment by fabiand** [comment]
/hold

I'm just wondering: What changed that this flag is now suddenly required?
Also: invtsc often comes with a frequency - why does the frequency not have to be set here?

**Comment by iholder-redhat** [comment]
Hey Fabian,

> I'm just wondering: What changed that this flag is now suddenly required?

Such VMs were always able to start and run, but not to migrate.
The implementation didn't change from the original PR here: https://github.com/kubevirt/kubevirt/pull/5742. I guess it was a hidden bug until now.

> Also: invtsc often comes with a frequency - why does the frequency not have to be set here?

As explained in the PR description:
```
In converter.go [1], we're adding a tsc frequency parameter if
vmi.Status.TopologyHints != nil && vmi.Status.TopologyHints.TSCFrequency != nil.

In virt-controller [2] we are checking if TopologyHints need to be added to the vmi by
calling TopologyHintsRequiredForVMI(). There, we are checking if "invtsc" feature
exists in vmi's spec (vmi.Spec.Domain.CPU.Features) with a "required" or "force"
policy. Only then we add the TopologyHints to the vmi, which will lead (in [1]) to adding
this parameter and pass it on to libvirt / QEMU.

[1] https://github.com/kubevirt/kubevirt/blob/v0.54.0/pkg/virt-launcher/virtwrap/converter/converter.go#L1691
[2] https://github.com/kubevirt/kubevirt/blob/v0.54.0/pkg/virt-controller/watch/vmi.go#L438
```

As written above, If "invtsc" is enabled, we add TopologyHints with desired tsc frequency to the VMI, then we're passing that frequency to libvirt.

**Comment by xpivarc** [comment]
I think what Fabian is suggesting is that the `invtsc` should not be required and rather we should make the VM not migratable. It should be the responsibility of the higher component to set appropriate features for specific VMs(Windows). Consider a use case where you are using or testing Kubevirt on top of virtualized hardware. You need to enable this flag with this PR and you might not have the possibility.

/cc @vladikr

**Comment by iholder-redhat** [comment]
> I think what Fabian is suggesting is that the `invtsc` should not be required and rather we should make the VM not migratable. It should be the responsibility of the higher component to set appropriate features for specific VMs(Windows).

Note that this PR enabled this feature only when `spec.Domain.Features.Hyperv.EVMCS != nil`, so it's being enabled only to specific windows VMs.

If we go with your logic, why do we enable "vmx" feature automatically in the exact same case? Why not also demand that the user does so explicitly?

> Consider a use case where you are using or testing Kubevirt on top of virtualized hardware. You need to enable this flag with this PR and you might not have the possibility.

We currently don't provide a way to set the tsc frequency manually. The only option is to set it automatically via enabling "inctsc" feature. Or am I missing something?

In any case, if we want to expose such an option in the future, it shouldn't be a problem since it can have precedence over the automatically set tsc frequency.

**Comment by iholder-redhat** [comment]
> I'm just wondering: What changed that this flag is now suddenly required?

Oh, forgot to link this: https://lists.gnu.org/archive/html/qemu-devel/2021-03/msg07116.html
It's a commit to QEMU from Match 2021 that required tsc frequency in order to migrate VMs with "vmx".

It was also linked in the [bugzilla page](https://bugzilla.redhat.com/show_bug.cgi?id=2100054).

**Comment by iholder-redhat** [comment]
> /hold
> 
> I'm just wondering: What changed that this flag is now suddenly required? Also: invtsc often comes with a frequency - why does the frequency not have to be set here?

We've sorted out things offline, but here's a quick summary:

> What changed that this flag is now suddenly required?

This [QEMU commit](https://gitlab.com/qemu-project/qemu/-/commit/561dbb41b1d752098249128d8462aaadc56fd15d) broke backward compatibility and now demands 'tsc-frequency=' parameter in order to migrate HyperV Reenlightenment VMs. We are affected since we bumped QEMU/libvirt versions.

> Also: invtsc often comes with a frequency - why does the frequency not have to be set here?

The frequency that's being set is the lowest tsc frequency on the cluster. Behind the scenes, every node sets this info in a label (through node-labeller) then we get the minimum frequency and add it to VMI's TopologyHints if needed (before this PR - for VMIs with invtsc feature. After this PR - also for HyperV Reenlightenment VMs).

/hold cancel

**Comment by iholder-redhat** [comment]
PTAL @vladikr @xpivarc

The PR description is now updated. Here's the updated "Fix" part:
```
As explained above, TopologyHints are added to a vmi if it has the invtsc feature and this
will eventually cause us to provide 'tsc-frequency=' flag for QEMU [1].

After this PR, TopologyHints will be added if either invtsc is enabled for the vmi (as before)
or if the VMI is using HyperV Reenlightenment feature.
```

In other word, `invtsc` is not being added, but instead the 'tsc-frequency=' parameter is provided to QEMU if HyperV Reenlightenment is enabled.

**Comment by fabiand** [comment]
FWIW

The Fix seciton includes the sentence

> After this PR, TopologyHints will be added if either invtsc is enabled for the vmi (as before) or if the VMI is using HyperV Reenlightenment feature.

What I am missing form this description is what this means for the VMI API, thus for the user.
Is the following statement correct?

> After this PR, If a user uses the HyperV Reenlightenment feature today and runs into this bug, then after this change, the same VM spec will migrate, because on the lower levels the tsc frequency will be set.

**Comment by iholder-redhat** [comment]
> FWIW
> 
> The Fix seciton includes the sentence
> 
> > After this PR, TopologyHints will be added if either invtsc is enabled for the vmi (as before) or if the VMI is using HyperV Reenlightenment feature.
> 
> What I am missing form this description is what this means for the VMI API, thus for the user. Is the following statement correct?
> 
> > After this PR, If a user uses the HyperV Reenlightenment feature today and runs into this bug, then after this change, the same VM spec will migrate, because on the lower levels the tsc frequency will be set.

Yes, this is correct.
After this merges, there wouldn't be a need of any user intervention in order for such VMs to migrate.

**Comment by vladikr** [comment]
I wouldn't mention `invtsc` at all. As we've discussed, you only want to re-use the same mechanism that finds the lowest TSC freq. on the cluster and add it when the hyper-v enlightenment feature is enabled.

**Comment by iholder-redhat** [comment]
> Workaround



> I wouldn't mention `invtsc` at all. As we've discussed, you only want to re-use the same mechanism that finds the lowest TSC freq. on the cluster and add it when the hyper-v enlightenment feature is enabled.

I only mention it under `Workaround` section, as it's relevant there.
Implementation-wise, as you say, invtsc is not being applied, but instead the lowest TSC mechanism is being re-used.

**Review by stu-gott — COMMENTED** [review]
/approve

Code looks good @iholder-redhat. We discussed this offline, but repeating it here as it's worth noting for the record.

I have a concern as to the upgrade path.

This change came because qemu now requires tsc_frequency to be set. If a VM was running an older version of KubeVirt/qemu it will not necessarily have the flag set. If it migrates to a KubeVirt version past when this PR lands, it will suddenly have this value set. There's a decent chance the freq that's arbitrarily chosen will happen to match the implicit tsc_frequency from the previous host (migration source), but there's also a chance of a mismatch. To be clear this is no worse than previous behavior where the implicit tsc_freq inherited from the host would also just silently mismatch. In this unlucky case the VM will almost certainly lock up or crash shortly after migration. This gotcha should likely be noted somewhere: it's probably best to do a one-time restart so that Operating System timers that depend on tsc frequency can be reset.

**Comment by stu-gott** [comment]
retest-required

**Review by xpivarc — CHANGES_REQUESTED** [review]
This PR needs rebase.

Some suggestions are in the comments.

File: pkg/virt-api/webhooks/hyperv.go
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/hyperv.go:180]: No, we can't :)

From this point on, all of the checks are regarding EVMCS. In the upper check, we general HyperV dependencies, not only EVMCS ones.

You can look at the original condition and see that no logic is changed. This commit is 100% a refactoring that has no effect on actual behavior.

File: pkg/virt-api/webhooks/hyperv.go
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/hyperv.go:184]: No, it isn't.
Again, this is a refactoring commit, no logic is changed, just cleaned up (this was very messy in its earlier form imo)

File: pkg/virt-api/webhooks/hyperv.go
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/hyperv.go:187]: No, same answer as above

File: pkg/virt-api/webhooks/hyperv.go
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/hyperv.go:193]: Right! :)
thanks for catching that!

File: pkg/virt-api/webhooks/hyperv.go
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/hyperv.go:198]: nope

**Comment by iholder-redhat** [comment]
@xpivarc and I talked offline, seems there was a misunderstanding here.
The PR is ready for review.

/hold cancel

File: pkg/virt-api/webhooks/hyperv.go
Comment (xpivarc) [reply pkg/virt-api/webhooks/hyperv.go:181]: What I meant by this is: Why should we actually call `getHypervFeatureDependencies` if `spec.Domain.Features == nil || spec.Domain.Features.Hyperv == nil`. But I see it would just return immediately. Moving this to the top would make it more obvious imo.

File: pkg/virt-api/webhooks/hyperv.go
Comment (xpivarc) [reply pkg/virt-api/webhooks/hyperv.go:185]: I don't agree. This commit creates `getEVMCSDependency` which returns an array. In the previous version of this PR, this was required but now it is not. Therefore now you have redundant `for` loops.

File: pkg/virt-controller/watch/vmi_test.go
Comment (xpivarc) [inline pkg/virt-controller/watch/vmi_test.go:2768]: Don't you want to actually see what was updated?

File: tests/migration_test.go
Comment (xpivarc) [inline tests/migration_test.go:4230]: Do you actually require it?

File: pkg/virt-api/webhooks/hyperv.go
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/hyperv.go:185]: What if we add more dependencies in the future? This sets the ground for it and is more generic.
In any case there is no change to behavior.

File: pkg/virt-api/webhooks/hyperv.go
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/hyperv.go:181]: > But I see it would just return immediately

I don't see a point in checking this if it's already being checked in `getHypervFeatureDependencies()`. I call the function, it will tell me if there are dependencies or not. I don't need to even know what it's checking inside and I see no reason to explicitly check it.

File: pkg/virt-api/webhooks/hyperv.go
Comment (xpivarc) [reply pkg/virt-api/webhooks/hyperv.go:185]: We did not need to add it until now and it is pretty straightforward to do in the future if we need it.

File: pkg/virt-api/webhooks/hyperv.go
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/hyperv.go:181]: But these are different checks..
Currently we:
- call getHypervFeatureDependencies()
  - Which inside checks if `spec.Domain.Features == nil || spec.Domain.Features.Hyperv == nil` and if so it returns
- check if EVMCS is enabled (by checking all 3 conditions, e.g. `Features == nil || Features.Hyperv != nil || Features.Hyperv.EVMCS == nil`

The check for EVMCS cannot be done at the top since even if it's not enabled we should take care of getHypervFeatureDependencies().

We can instead:
- Check for hypev features
- Call getHypervFeatureDependencies() (which, again, would do the same check again internally)
- Check for EVMCS (with exactly the same conditions as now).

I don't think it's more obvious, quite the contrary. I want to check for dependencies, and get an empty feature slice if there are none. I don't care how the function checks it and I don't need to re-check the same condition it checks. Another point is that if someday these conditions are changed we now have to change them in two different places. I honestly don't think it's any better.

But we can agree to disagree and move on.

Comment (vladikr) [reply pkg/virt-api/webhooks/hyperv.go:181]: We're trying to optimize a single if statement ... ;)
If we really want to optimize it, let's do it in a follow-up. 

Btw, `getHypervFeatureDependencies` can't even return nil.

**Review by vladikr — COMMENTED** [review]
Looks great. The main idea is implemented and well tested. 
/lgtm
/hold

@xpivarc please unhold if your concerns are addressed.

**Comment by xpivarc** [comment]
/hold
@iholder-redhat I think there are linter issues. Can you have a look?

**Comment by iholder-redhat** [comment]
> /hold @iholder-redhat I think there are linter issues. Can you have a look?

Indeed a silly linter error.
Can you please re-LGTM?

**Pull request merged**
